osgp/platform/osgp-adapter-ws-smartmetering/src/main/java/org/opensmartgridplatform/adapter/ws/smartmetering/endpoints/SmartMeteringConfigurationEndpoint.java

```diff
@@ -1586,4 +1609,133 @@ public SetRandomisationSettingsResponse getSetRandomisationSettingsResponse(
+
+      final ResponseData responseData =
+          this.responseDataService.dequeue(
+              request.getCorrelationUid(), ComponentType.WS_SMART_METERING);
+
+      this.throwExceptionIfResultNotOk(responseData, "getting keys");
+
+      response.setResult(OsgpResultType.fromValue(responseData.getResultType().getValue()));
+
+      if (responseData.getMessageData() != null) {
+        final org.opensmartgridplatform.domain.core.valueobjects.smartmetering.GetKeysResponse
+            getKeysResponse =
+                (org.opensmartgridplatform.domain.core.valueobjects.smartmetering.GetKeysResponse)
+                    responseData.getMessageData();
+
+        final List<GetKeysResponseData> keysEncryptedWithGxfKey =
+            this.configurationMapper.mapAsList(
+                getKeysResponse.getKeys(), GetKeysResponseData.class);
+
+        final List<GetKeysResponseData> keysEncryptedForApplication =
+            this.reencryptKeysInGetKeysResponse(
+                keysEncryptedWithGxfKey, responseData.getOrganisationIdentification());
+
+        response.getGetKeysResponseData().addAll(keysEncryptedForApplication);
+      } else {
+        log.info("Get keys response is null");
+      }
+    } catch (final Exception e) {
+      this.handleException(e);
+    }
+    return response;
+  }
+
+  private List<GetKeysResponseData> reencryptKeysInGetKeysResponse(
+      final List<GetKeysResponseData> keysEncryptedWithGxfKey,
+      final String organisationIdentification)
+      throws OsgpException {
+    final RsaEncrypter applicationRsaEncrypter =
+        this.getRsaEncrypterForApplication(organisationIdentification);
+
+    return keysEncryptedWithGxfKey.stream()
+        .map(key -> this.reencryptKeyInGetKeysResponse(key, applicationRsaEncrypter))
+        .collect(Collectors.toList());
+  }
+
+  private GetKeysResponseData reencryptKeyInGetKeysResponse(
+      final GetKeysResponseData data, final RsaEncrypter applicationRsaEncrypter) {
+
+    final byte[] decryptedKey = this.gxfRsaDecrypter.decrypt(data.getSecretValue());
+
+    data.setSecretValue(applicationRsaEncrypter.encrypt(decryptedKey));
+
+    return data;
+  }
+
+  private RsaEncrypter getRsaEncrypterForApplication(final String organisationIdentification)
+      throws OsgpException {
+    final Optional<ApplicationKeyConfiguration> applicationKeyConfiguration =
+        this.applicationKeyConfigurationRepository.findById(
+            new ApplicationKeyLookupKey(
+                organisationIdentification, ApplicationConstants.APPLICATION_NAME));
+
+    if (!applicationKeyConfiguration.isPresent()) {
+      throw new OsgpException(
+          ComponentType.WS_SMART_METERING,
+          "No public key found for application "
+              + ApplicationConstants.APPLICATION_NAME
+              + " and organisation "
+              + organisationIdentification);
+    }
+
+    final String publicKeyLocation = applicationKeyConfiguration.get().getPublicKeyLocation();
```

Comment:
If you find yourself introducing optional as variable, calling `isPresent` and later `get` the issues with unexpected exceptions on calling get with an empty optional are avoided, but there probably is a more elegant way to deal with the code scenario.
One thing that often looks a bit off to me is the name of the optional: do you call it something ending with Optional for instance, or use a name that actually fits better with the value inside the optional, like in the code above.

You could consider avoiding all this by using some of the richer parts of the Optional API, like:
```java
    final ApplicationKeyConfiguration applicationKeyConfiguration =
        this.applicationKeyConfigurationRepository
            .findById(
                new ApplicationKeyLookupKey(
                    organisationIdentification, ApplicationConstants.APPLICATION_NAME))
            .orElseThrow(
                () ->
                    new OsgpException(
                        ComponentType.WS_SMART_METERING,
                        "No public key found for application "
                            + ApplicationConstants.APPLICATION_NAME
                            + " and organisation "
                            + organisationIdentification));

    final String publicKeyLocation = applicationKeyConfiguration.getPublicKeyLocation();
```

Reply:
Updated the code as suggested.